source/_posts/2015-11-16-immutable-strings-in-ruby-2-dot-3.markdown

```diff
@@ -0,0 +1,279 @@
+      from replace.rb:2:in `<main>'
+
+As you can see a subtle change is made to the error output, now we have exact
+information about were the string was created.
+
+### Same value literals points to the same object
+
+Same value literals points to the same object, i.e. `a` and `a` points to the
+same object. In Ruby we can compare two objects in different ways. Let's run a
+small test with and without frozen strings to see what's changed.
+
+{% codeblock lang:ruby %}
+ puts 'a' == 'a'
+ puts 'a'.equal? 'a'
+ puts String.new('a').equal? String.new('a')
+{% endcodeblock %}
+
+Without frozen strings for `'a'.equal? 'a'` the interpreter creates two objects
+with the value `a`, hence `equal?` will return `false`. With immutable strings
+this is no longer true, in this case the interpreter will use a reference to the
+same object, then as each reference points to the same object `equal?` will
+return `true`.
+
+|Operation|Mutable|Immutable|
+|-|-|-|
+|'a' == 'a'|true|true|
+|'a'.equal? 'a'|false|true|
+|String.new('a').equal? String.new('a')|false|false|
+
+So if you are planning to upgrade your project to frozen strings keep this behavior
+in mind and check for existent comparisons.
+
+
+### Thread safety
+
+Let's see what wikipedia says about [thread safety][4]:
+
+> A piece of code is thread-safe if it only manipulates shared data structures
+> in a manner that guarantees safe execution by multiple threads at the same time.
+
+When dealing with objects we have two basic operations: read and write.
+Concurrent reading is inherently thread safe, as we're not changing the state
+of the object we're reading. On the other hand we have write operations, in this
+case the object state change thus it must be synchronized. When an object state
+can change at any moment, we need to synchronize both operations. This happens
+because writing on the wrong moment can lead objects to be on an invalid state
+and reading can get us an invalid state.
+
+With all this in mind, we can say that string literals (and immutable objects in general)
+are thread safe. Its state can't change as we can't change its value and if we try to,
+an error will be thrown. Therefore race conditions are gone.
+
+Certainly forgetting about race conditions in certain scenarios is great. But
+(there's always a but) bear in mind that string literals are thread safe, not
+its references!
+
+{% codeblock lang:ruby %}
+class Foo
+  attr_accesor :literal
+
+  def initialize(literal)
+    @literal = literal
+  end
+end
+{% endcodeblock %}
+
+Picture yourself working with a `Foo` class, what happens when we pass a `Foo`
+object to several threads, some of them reading from `literal` attribute and
+another ones writing to?
+
+In this case despite we are using string literals, we aren't safe when it comes
+to thread safety. Actually we're holding a reference to a literal, nothing stop
+us from changing that reference for another literal.
```

Comment:
Not sure I understand properly what you mean, but can you describe with an example what you're saying that is not thread safe?